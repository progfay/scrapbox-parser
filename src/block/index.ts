import { convertToCodeBlock } from "./CodeBlock";
import { convertToLine } from "./Line";
import { convertToTable } from "./Table";
import { convertToTitle } from "./Title";

import type { CodeBlock } from "./CodeBlock";
import type { Line } from "./Line";
import type { Pack } from "./Pack";
import type { Table } from "./Table";
import type { Title } from "./Title";

/**
 * Scrapbox block type
 */
export type Block = Title | CodeBlock | Table | Line;

export const convertToBlock = (pack: Pack): Block => {
	switch (pack.type) {
		case "title":
			return convertToTitle(pack);

		case "codeBlock":
			return convertToCodeBlock(pack);

		case "table":
			return convertToTable(pack);

		case "line":
			return convertToLine(pack);
	}
};
