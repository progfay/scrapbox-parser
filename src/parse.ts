import { BlockType, convertToBlock } from './block'
import { BlockComponentType, convertToBlockComponents } from './block/BlockComponent'
import { PackedBlockComponentType, packBlockComponents } from './block/PackedBlockComponent'

type PageType = {
  title: string
  blocks: BlockType[]
}

export const convertToBlocks = (blockComponents: BlockComponentType[]): BlockType[] => {
  const packedBlockComponents: PackedBlockComponentType[] = packBlockComponents(blockComponents)
  return packedBlockComponents.map(convertToBlock)
}

export const parse = (input: string): PageType => {
  const blockComponents: BlockComponentType[] = convertToBlockComponents(input.trim())

  const [firstBlock, ...body] = blockComponents
  const title = firstBlock && firstBlock.text ? firstBlock.text : 'Untitled'
  const blocks = convertToBlocks(body)

  return { title, blocks }
}
