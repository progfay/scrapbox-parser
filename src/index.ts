import { parse } from './parse'

export default parse
export { BlockType } from './block'
export { CodeBlockType } from './block/CodeBlock'
export { TableType } from './block/Table'
export { LineType } from './block/Line'
export { LineNodeType } from './block/node'
export { QuoteNodeType } from './block/node/QuoteNode'
export { HelpfeelNodeType } from './block/node/HelpfeelNode'
export { StrongImageNodeType } from './block/node/StrongImageNode'
export { StrongIconNodeType } from './block/node/StrongIconNode'
export { StrongNodeType } from './block/node/StrongNode'
export { FormulaNodeType } from './block/node/FormulaNode'
export { DecorationNodeType, DecorationType } from './block/node/DecorationNode'
export { CodeNodeType } from './block/node/CodeNode'
export { BlankNodeType } from './block/node/BlankNode'
export { UrlNodeType } from './block/node/UrlNode'
export { GoogleMapNodeType } from './block/node/GoogleMapNode'
export { InternalLinkNodeType } from './block/node/InternalLinkNode'
export { IconNodeType } from './block/node/IconNode'
export { HashTagNodeType } from './block/node/HashTagNode'
export { PlainNodeType } from './block/node/PlainNode'
