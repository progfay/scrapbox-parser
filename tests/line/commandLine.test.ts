import { describe, it, expect } from "vitest";
import { parse } from "../../src";

describe("commandLine", () => {
  it("Simple command with $", () => {
    expect(parse("$ command", { hasTitle: false })).toMatchSnapshot();
  });

  it("Simple command with %", () => {
    expect(parse("% command", { hasTitle: false })).toMatchSnapshot();
  });

  it("`$` is not command", () => {
    expect(parse("$", { hasTitle: false })).toMatchSnapshot();
  });

  it("`$ ` is not command", () => {
    expect(parse("$ ", { hasTitle: false })).toMatchSnapshot();
  });

  it("`$s` is not command", () => {
    expect(parse("$not command", { hasTitle: false })).toMatchSnapshot();
  });
});
