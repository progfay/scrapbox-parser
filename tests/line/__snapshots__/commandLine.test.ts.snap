// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`commandLine \`$ \` is not command 1`] = `
Array [
  Object {
    "indent": 0,
    "nodes": Array [
      Object {
        "raw": "$ ",
        "text": "$ ",
        "type": "plain",
      },
    ],
    "type": "line",
  },
]
`;

exports[`commandLine \`$\` is not command 1`] = `
Array [
  Object {
    "indent": 0,
    "nodes": Array [
      Object {
        "raw": "$",
        "text": "$",
        "type": "plain",
      },
    ],
    "type": "line",
  },
]
`;

exports[`commandLine \`$s\` is not command 1`] = `
Array [
  Object {
    "indent": 0,
    "nodes": Array [
      Object {
        "raw": "$not command",
        "text": "$not command",
        "type": "plain",
      },
    ],
    "type": "line",
  },
]
`;

exports[`commandLine Simple command with $ 1`] = `
Array [
  Object {
    "indent": 0,
    "nodes": Array [
      Object {
        "raw": "$ command",
        "symbol": "$",
        "text": "command",
        "type": "commandLine",
      },
    ],
    "type": "line",
  },
]
`;

exports[`commandLine Simple command with % 1`] = `
Array [
  Object {
    "indent": 0,
    "nodes": Array [
      Object {
        "raw": "% command",
        "symbol": "%",
        "text": "command",
        "type": "commandLine",
      },
    ],
    "type": "line",
  },
]
`;
